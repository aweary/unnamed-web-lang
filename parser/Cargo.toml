[package]
name = "parser"
version = "0.1.0"
authors = ["Brandon Dail <brandondail@fb.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
graphql-parser = "0.2.3"
salsa = "0.14.1"
lexer = { path = "../lexer" }
syntax = { path = "../syntax" }
diagnostics = { path = "../diagnostics" }
source = { path = "../source" }
insta = { version = "0.16", features = ["glob"] }
log = "0.4.8"