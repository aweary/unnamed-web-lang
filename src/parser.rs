use ast::*;
use lexer::Lexer;
use token::*;

pub struct Parser {
    pub lexer: Lexer,
    // state: ParserState,
}

impl Parser {
    pub fn module(source: String) -> Self {
        let lexer = Lexer::new(source);
        // let state = ParserState::new();
        Parser { lexer }
    }

    fn peek(&mut self) -> Option<&Token> {
        self.lexer.peek_token()
    }

    pub fn skip_until<F>(&mut self, predicate: F)
    where
        F: Fn(Token) -> bool,
    {
        loop {
            match self.lexer.next_token() {
                Some(token) => {
                    if predicate(token) {
                        return;
                    }
                }
                None => return,
            }
        }
    }

    pub fn skip_assert(&mut self, token: Token) {
        assert_eq!(Some(token), self.lexer.next_token())
    }

    fn read_param_list(&mut self) -> Option<Vec<Param>> {
        match self.lexer.next_token() {
            // If the next token is an opening curly bracket, there
            // are no arguments and we assume this component takes no input
            Some(Token::LCurlyBracket) => None,
            // If the next token is an opening paren, there is
            // an argument list.
            Some(Token::LParen) => {
                let mut params: Vec<Param> = vec![];
                loop {
                    let name = self.read_ident();
                    // TODO this should be an unexpected token error
                    assert_eq!(Some(Token::Colon), self.lexer.next_token());
                    let kind = self.read_ident();
                    params.push(Param::new(name, kind));
                    match self.lexer.next_token() {
                        Some(Token::RParen) => {
                            break;
                            // End of the param list
                        }
                        Some(Token::Comma) => {
                            continue;
                            // There's another paramter
                        }
                        // Anything else is an unexpected token error
                        _ => unreachable!(),
                    }
                }
                Some(params)
            }
            _ => unreachable!("Unexpected token"),
        }
        // Next token should be the opening paren for the argument list
    }

    fn read_ident(&mut self) -> String {
        match self.lexer.next_token() {
            Some(Token::Ident(name)) => name,
            unknown @ _ => unreachable!("Expected identifier, found: {:?}", unknown),
        }
    }

    // Assumes that the current token is Reserved(Component)
    fn read_component_declaration(&mut self) -> Node {
        // Next token should be the identifier for the component definition
        let name = self.read_ident();
        // Read list of all params
        let params = self.read_param_list();
        // Everything from here until the closing bracket is part of the
        // component definition
        let body = self.read_component_body();
        let decl = ComponentDefinition::new(name, params, body);
        Node::ComponentDefinition(decl)
    }

    // A component is composed of a set of expressions/statements.
    fn read_component_body(&mut self) -> Vec<Expression> {
        let mut exprs = vec![];
        loop {
            match self.lexer.next_token() {
                // Statements starting with '<' are assumed to be
                // the beginning of JSX statements.
                Some(Token::LessThan) => {
                    let expr = Expression::JSXElement(self.read_jsx_element());
                    exprs.push(expr);
                    return exprs;
                }
                _ => {
                    continue;
                }
            }
        }
    }

    /**
     * read_expr
     * Reads any expresison, which is something that can evaluate to some value
     */
    fn read_expr(&mut self) -> Expression {
        self.skip_until(|token| token == Token::RCurlyBracket);
        Expression::NumericLiteral(42)
    }

    fn read_jsx_attributes(&mut self) -> Option<Vec<JSXAttribute>> {
        let mut attrs: Vec<JSXAttribute> = vec![];
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let attr = self.read_jsx_attribute();
                    attrs.push(attr);
                }
                Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                    break;
                }
                _ => {
                    unreachable!("Unknown token: {:?}", self.peek());
                }
            }
        }
        println!("read_jsx_attributes: {:?}", attrs);
        if attrs.is_empty() {
            None
        } else {
            Some(attrs)
        }
    }

    fn is_jsx_element_self_closing(&mut self) -> bool {
        match self.peek() {
            Some(&Token::GreaterThan) => {
                self.skip_assert(Token::GreaterThan);
                false
            }
            Some(&Token::ForwardSlash) => {
                self.skip_assert(Token::ForwardSlash);
                self.skip_assert(Token::GreaterThan);
                true
            }
            _ => {
                unreachable!();
            }
        }
    }

    fn read_jsx_element(&mut self) -> JSXElement {
        let tag_name = self.read_ident();
        let attributes = self.read_jsx_attributes();
        let self_closing = self.is_jsx_element_self_closing();
        let opening_element = JSXOpeningElement::new(tag_name.clone(), attributes, self_closing);
        println!("opening_element {:?}", opening_element);
        let (children, closing_element) = if self_closing {
            (None, None)
        } else {
            let children = self.read_jsx_children();
            let closing_element = self.read_jsx_closing_element();
            println!("closing_element {:?}", closing_element);
            println!("children {:?}", children);
            (Some(children), Some(closing_element))
        };
        let element = JSXElement::new(opening_element, closing_element, children);
        println!("element: {:#?}", element);
        element
    }

    fn read_jsx_closing_element(&mut self) -> JSXClosingElement {
        // Closing tag
        // self.skip_assert(Token::LessThan);
        // self.skip_assert(Token::ForwardSlash);
        let closing_tag_name = self.read_ident();
        // TODO parsing error
        // assert_eq!(tag_name, closing_tag_name);
        self.skip_assert(Token::GreaterThan);
        // TODO actually parse the closing tag
        JSXClosingElement::new(closing_tag_name)
    }

    fn read_jsx_text(&mut self) -> Box<JSXChildren> {
        let mut text = String::new();
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let word = self.read_ident();
                    // TODO respect whitespace when lexing instead of this hack
                    text = text + " " + &word;
                }
                Some(&Token::LessThan) | Some(&Token::LCurlyBracket) => {
                    break;
                }
                _ => {
                    unreachable!();
                }
            }
        }
        println!("read_jsx_text: {:?}", text);
        Box::new(JSXChildren::JSXText(text))
    }

    fn read_jsx_children(&mut self) -> Vec<Box<JSXChildren>> {
        let mut children: Vec<Box<JSXChildren>> = vec![];
        loop {
            match self.peek() {
                // JSXText
                Some(&Token::Ident(_)) => {
                    children.push(self.read_jsx_text());
                },
                Some(&Token::LessThan) => {
                    // A '<' token could be the start of another element
                    // or the closing element for the enclosing element
                    self.skip_assert(Token::LessThan);
                    match self.peek() {
                        Some(&Token::Ident(_)) => {
                            // An identifier means the start of a new element
                            let element = self.read_jsx_element();
                            children.push(Box::new(JSXChildren::JSXElement(element)));
                        }
                        Some(&Token::ForwardSlash) => {
                            self.skip_assert(Token::ForwardSlash);
                            return children;
                            // A '/' indicates a closing element
                        }
                        _ => {
                            unreachable!();
                        }
                    }
                }
                _ => break,
            }
        }
        children
        // let child = match token {
        //     // TODO JSX parsers respect the whitespace inside tags. Need to output tokens
        //     Some(Token::Ident(word)) => {
        //         JSXChildren::JSXText(word)
        //     }
        //     Some(Token::LessThan) => {
        //         let element = self.read_jsx_element();
        //         JSXChildren::JSXElement(element)
        //     }
        //     // TODO handle JSXExpressionContainer
        //     token @ _ => {
        //         unreachable!("Unexpected token: {:?}", token);
        //     }
        // };
        // Box::new(child)
    }

    /**
     * Reads the expression part of a JSX attribute.
     */
    fn read_jsx_attribute_value(&mut self) -> JSXAttributeValue {
        // An explicit assignment. The assigned value can
        // either be a string literal or an identifier in curly brackets
        match self.lexer.next_token() {
            Some(Token::String(literal)) => JSXAttributeValue::StringLiteral(literal),
            Some(Token::LCurlyBracket) => {
                let expr = self.read_expr();
                JSXAttributeValue::Expression(expr)
            }
            token @ _ => unreachable!("Unknown token: {:?}", token),
        }
    }

    /**
     * read_jsx_attr
     *
     * A JSX Attribute comes in two forms:
     *
     * 1. An explicit attribute where the value is passed
     *    in inline (<div id="1" />)
     * 2. An implicit attribute where we assume the attribute
     *    name is a reference to some identifier
     *    (<div id>)
     */
    fn read_jsx_attribute(&mut self) -> JSXAttribute {
        let name = self.read_ident();

        let foo = match self.peek() {
            Some(&Token::Assign) => {
                self.skip_assert(Token::Assign);
                let attribute_value = self.read_jsx_attribute_value();
                JSXAttribute::new(name, attribute_value)
            }
            Some(&Token::Ident(_)) | Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                let attr_ident = Identifier::new(name.clone());
                let attr_ident_expr = Expression::Identifer(attr_ident);
                let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
                JSXAttribute::new(name, attr_value)
            }
            _ => {
                unreachable!();
            }
        };
        foo
        // println!("read_jsx_attribute:name {:?}", name);
        // match self.lexer.next_token() {
        //     // Explicit assignment
        //     Some(Token::Assign) => {
        //         let attr_value = self.read_jsx_attribute_value();
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     // Shorthand form
        //     Some(Token::Ident(_)) | Some(Token::ForwardSlash) | Some(Token::GreaterThan) => {
        //         println!("shorthand form, creating the identifier and stuff");
        //         let attr_ident = Identifier::new(name.clone());
        //         let attr_ident_expr = Expression::Identifer(attr_ident);
        //         let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     token @ _ => unreachable!("Unknown character: {:?}", token),
        // }
    }

    fn read_type_def(&mut self) -> Node {
        // Next token is the identifier for the type definition.
        let name = self.read_ident();
        let def = TypeDefinition::new(name);
        // Just read everything until we reach the end for now...
        loop {
            if self.lexer.next_token() == Some(Token::RCurlyBracket) {
                break;
            }
        }
        Node::TypeDefinition(def)
    }

    /**
     * Inside a module, top level nodes are currently restricted to component
     * and function definitions. This means you cannot have other expressions
     * that are executed when the module is evalulated. All components and
     * functions must be lazily evaluated.
     */
    pub fn next_node(&mut self) -> Node {
        let token = self.lexer.next_token().unwrap();
        match token {
            Token::Reserved(ReservedWord::Component) => self.read_component_declaration(),
            Token::Reserved(ReservedWord::Type) => self.read_type_def(),
            _ => Node::Text,
        }
    }
}
