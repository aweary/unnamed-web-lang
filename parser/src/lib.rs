#![warn(clippy::all, clippy::pedantic)]
#![allow(clippy::must_use_candidate)]

mod parser;

pub use crate::parser::*;

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use insta::assert_json_snapshot;
    use syntax::ast;

    fn parse_module(source: &str) -> ast::Module {
        let mut parser = Parser::new(source);
        parser.parse_module().unwrap()
    }

    #[test]
    fn fixture_function_simple() {
        let input = include_str!("./inputs/function_simple.dan");
        assert_json_snapshot!(parse_module(input));
    }

    #[test]
    fn fixture_control_flow() {
        let input = include_str!("./inputs/control_flow.dan");
        assert_json_snapshot!(parse_module(input));
    }

    #[test]
    fn fixture_trailing_closure_control_flow() {
        let input = include_str!("./inputs/trailing_closures.dan");
        assert_json_snapshot!(parse_module(input));
    }
}
