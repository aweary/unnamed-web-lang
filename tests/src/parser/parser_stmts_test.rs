#[cfg(test)]
mod tests {
    use parser::Parser;
    use std::fs;
    use insta::{glob, assert_json_snapshot};
    #[test]
    // Easy but very noisy testing. Hard to debug when something fails.
    // Good for now, but more granular tests in the future will be great.
    fn parser_tests() {
        glob!("fixtures/*.dan", |path| {
            let input = fs::read_to_string(path).unwrap();
            let mut parser = Parser::new(&input);
            let ast = parser.parse_module().unwrap();
            assert_json_snapshot!(ast);
        });
    }
}