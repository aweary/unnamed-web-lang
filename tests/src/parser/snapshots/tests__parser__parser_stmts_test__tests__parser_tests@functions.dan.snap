---
source: tests/src/parser/parser_stmts_test.rs
expression: ast
input_file: tests/src/parser/fixtures/functions.dan
---
{
  "items": [
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "empty"
            },
            "span": {
              "start": 3,
              "end": 8
            }
          },
          "params": "Empty",
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 11,
              "end": 13
            }
          },
          "return_ty": null,
          "is_async": false,
          "generics": null,
          "span": {
            "start": 0,
            "end": 13
          }
        }
      },
      "span": {
        "start": 0,
        "end": 13
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "empty_annotated"
            },
            "span": {
              "start": 17,
              "end": 32
            }
          },
          "params": "Empty",
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 42,
              "end": 44
            }
          },
          "return_ty": {
            "Unit": {
              "start": 37,
              "end": 41
            }
          },
          "is_async": false,
          "generics": null,
          "span": {
            "start": 14,
            "end": 44
          }
        }
      },
      "span": {
        "start": 14,
        "end": 44
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "one_arg"
            },
            "span": {
              "start": 48,
              "end": 55
            }
          },
          "params": {
            "Single": {
              "local": {
                "Ident": [
                  {
                    "symbol": {
                      "unstable_source": "a"
                    },
                    "span": {
                      "start": 56,
                      "end": 57
                    }
                  },
                  {
                    "start": 56,
                    "end": 57
                  }
                ]
              },
              "ty": null,
              "id": 0,
              "span": {
                "start": 56,
                "end": 57
              }
            }
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 59,
              "end": 61
            }
          },
          "return_ty": null,
          "is_async": false,
          "generics": null,
          "span": {
            "start": 45,
            "end": 61
          }
        }
      },
      "span": {
        "start": 45,
        "end": 61
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "on_arg_annotated"
            },
            "span": {
              "start": 65,
              "end": 81
            }
          },
          "params": {
            "Single": {
              "local": {
                "Ident": [
                  {
                    "symbol": {
                      "unstable_source": "a"
                    },
                    "span": {
                      "start": 82,
                      "end": 83
                    }
                  },
                  {
                    "start": 82,
                    "end": 83
                  }
                ]
              },
              "ty": {
                "String": {
                  "start": 85,
                  "end": 91
                }
              },
              "id": 0,
              "span": {
                "start": 82,
                "end": 91
              }
            }
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 100,
              "end": 102
            }
          },
          "return_ty": {
            "Unit": {
              "start": 95,
              "end": 99
            }
          },
          "is_async": false,
          "generics": null,
          "span": {
            "start": 62,
            "end": 102
          }
        }
      },
      "span": {
        "start": 62,
        "end": 102
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "on_arg_annotated"
            },
            "span": {
              "start": 106,
              "end": 122
            }
          },
          "params": {
            "Single": {
              "local": {
                "Ident": [
                  {
                    "symbol": {
                      "unstable_source": "a"
                    },
                    "span": {
                      "start": 123,
                      "end": 124
                    }
                  },
                  {
                    "start": 123,
                    "end": 124
                  }
                ]
              },
              "ty": {
                "String": {
                  "start": 126,
                  "end": 132
                }
              },
              "id": 0,
              "span": {
                "start": 123,
                "end": 132
              }
            }
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 134,
              "end": 136
            }
          },
          "return_ty": null,
          "is_async": false,
          "generics": null,
          "span": {
            "start": 103,
            "end": 136
          }
        }
      },
      "span": {
        "start": 103,
        "end": 136
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "many_args"
            },
            "span": {
              "start": 140,
              "end": 149
            }
          },
          "params": {
            "Multi": [
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "a"
                      },
                      "span": {
                        "start": 150,
                        "end": 151
                      }
                    },
                    {
                      "start": 150,
                      "end": 151
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 150,
                  "end": 151
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "b"
                      },
                      "span": {
                        "start": 153,
                        "end": 154
                      }
                    },
                    {
                      "start": 153,
                      "end": 154
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 153,
                  "end": 154
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "c"
                      },
                      "span": {
                        "start": 156,
                        "end": 157
                      }
                    },
                    {
                      "start": 156,
                      "end": 157
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 156,
                  "end": 157
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "d"
                      },
                      "span": {
                        "start": 159,
                        "end": 160
                      }
                    },
                    {
                      "start": 159,
                      "end": 160
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 159,
                  "end": 160
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "e"
                      },
                      "span": {
                        "start": 162,
                        "end": 163
                      }
                    },
                    {
                      "start": 162,
                      "end": 163
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 162,
                  "end": 163
                }
              }
            ]
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 165,
              "end": 167
            }
          },
          "return_ty": null,
          "is_async": false,
          "generics": null,
          "span": {
            "start": 137,
            "end": 167
          }
        }
      },
      "span": {
        "start": 137,
        "end": 167
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "many_args_annotated"
            },
            "span": {
              "start": 171,
              "end": 190
            }
          },
          "params": {
            "Multi": [
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "a"
                      },
                      "span": {
                        "start": 191,
                        "end": 192
                      }
                    },
                    {
                      "start": 191,
                      "end": 192
                    }
                  ]
                },
                "ty": {
                  "String": {
                    "start": 194,
                    "end": 200
                  }
                },
                "id": 0,
                "span": {
                  "start": 191,
                  "end": 200
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "b"
                      },
                      "span": {
                        "start": 202,
                        "end": 203
                      }
                    },
                    {
                      "start": 202,
                      "end": 203
                    }
                  ]
                },
                "ty": {
                  "String": {
                    "start": 205,
                    "end": 211
                  }
                },
                "id": 0,
                "span": {
                  "start": 202,
                  "end": 211
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "c"
                      },
                      "span": {
                        "start": 213,
                        "end": 214
                      }
                    },
                    {
                      "start": 213,
                      "end": 214
                    }
                  ]
                },
                "ty": {
                  "String": {
                    "start": 216,
                    "end": 222
                  }
                },
                "id": 0,
                "span": {
                  "start": 213,
                  "end": 222
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "d"
                      },
                      "span": {
                        "start": 224,
                        "end": 225
                      }
                    },
                    {
                      "start": 224,
                      "end": 225
                    }
                  ]
                },
                "ty": {
                  "String": {
                    "start": 227,
                    "end": 233
                  }
                },
                "id": 0,
                "span": {
                  "start": 224,
                  "end": 233
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "e"
                      },
                      "span": {
                        "start": 235,
                        "end": 236
                      }
                    },
                    {
                      "start": 235,
                      "end": 236
                    }
                  ]
                },
                "ty": {
                  "String": {
                    "start": 238,
                    "end": 244
                  }
                },
                "id": 0,
                "span": {
                  "start": 235,
                  "end": 244
                }
              }
            ]
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 255,
              "end": 257
            }
          },
          "return_ty": {
            "Number": {
              "start": 248,
              "end": 254
            }
          },
          "is_async": false,
          "generics": null,
          "span": {
            "start": 168,
            "end": 257
          }
        }
      },
      "span": {
        "start": 168,
        "end": 257
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "many_args_line_breaks"
            },
            "span": {
              "start": 261,
              "end": 282
            }
          },
          "params": {
            "Multi": [
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "a"
                      },
                      "span": {
                        "start": 288,
                        "end": 289
                      }
                    },
                    {
                      "start": 288,
                      "end": 289
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 288,
                  "end": 289
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "b"
                      },
                      "span": {
                        "start": 295,
                        "end": 296
                      }
                    },
                    {
                      "start": 295,
                      "end": 296
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 295,
                  "end": 296
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "c"
                      },
                      "span": {
                        "start": 302,
                        "end": 303
                      }
                    },
                    {
                      "start": 302,
                      "end": 303
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 302,
                  "end": 303
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "d"
                      },
                      "span": {
                        "start": 309,
                        "end": 310
                      }
                    },
                    {
                      "start": 309,
                      "end": 310
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 309,
                  "end": 310
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "e"
                      },
                      "span": {
                        "start": 316,
                        "end": 317
                      }
                    },
                    {
                      "start": 316,
                      "end": 317
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 316,
                  "end": 317
                }
              }
            ]
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 321,
              "end": 323
            }
          },
          "return_ty": null,
          "is_async": false,
          "generics": null,
          "span": {
            "start": 258,
            "end": 323
          }
        }
      },
      "span": {
        "start": 258,
        "end": 323
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "mixed_annotations"
            },
            "span": {
              "start": 327,
              "end": 344
            }
          },
          "params": {
            "Multi": [
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "a"
                      },
                      "span": {
                        "start": 345,
                        "end": 346
                      }
                    },
                    {
                      "start": 345,
                      "end": 346
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 345,
                  "end": 346
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "b"
                      },
                      "span": {
                        "start": 348,
                        "end": 349
                      }
                    },
                    {
                      "start": 348,
                      "end": 349
                    }
                  ]
                },
                "ty": {
                  "String": {
                    "start": 351,
                    "end": 357
                  }
                },
                "id": 0,
                "span": {
                  "start": 348,
                  "end": 357
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "c"
                      },
                      "span": {
                        "start": 359,
                        "end": 360
                      }
                    },
                    {
                      "start": 359,
                      "end": 360
                    }
                  ]
                },
                "ty": null,
                "id": 0,
                "span": {
                  "start": 359,
                  "end": 360
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "d"
                      },
                      "span": {
                        "start": 362,
                        "end": 363
                      }
                    },
                    {
                      "start": 362,
                      "end": 363
                    }
                  ]
                },
                "ty": {
                  "Number": {
                    "start": 365,
                    "end": 371
                  }
                },
                "id": 0,
                "span": {
                  "start": 362,
                  "end": 371
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "e"
                      },
                      "span": {
                        "start": 373,
                        "end": 374
                      }
                    },
                    {
                      "start": 373,
                      "end": 374
                    }
                  ]
                },
                "ty": {
                  "String": {
                    "start": 376,
                    "end": 382
                  }
                },
                "id": 0,
                "span": {
                  "start": 373,
                  "end": 382
                }
              }
            ]
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 384,
              "end": 386
            }
          },
          "return_ty": null,
          "is_async": false,
          "generics": null,
          "span": {
            "start": 324,
            "end": 386
          }
        }
      },
      "span": {
        "start": 324,
        "end": 386
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "generic"
            },
            "span": {
              "start": 390,
              "end": 397
            }
          },
          "params": {
            "Single": {
              "local": {
                "Ident": [
                  {
                    "symbol": {
                      "unstable_source": "a"
                    },
                    "span": {
                      "start": 401,
                      "end": 402
                    }
                  },
                  {
                    "start": 401,
                    "end": 402
                  }
                ]
              },
              "ty": {
                "Reference": {
                  "name": {
                    "symbol": {
                      "unstable_source": "T"
                    },
                    "span": {
                      "start": 404,
                      "end": 405
                    }
                  },
                  "arguments": null,
                  "span": {
                    "start": 404,
                    "end": 405
                  }
                }
              },
              "id": 0,
              "span": {
                "start": 401,
                "end": 405
              }
            }
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 407,
              "end": 409
            }
          },
          "return_ty": null,
          "is_async": false,
          "generics": {
            "params": [
              {
                "symbol": {
                  "unstable_source": "T"
                },
                "span": {
                  "start": 398,
                  "end": 399
                }
              }
            ]
          },
          "span": {
            "start": 387,
            "end": 409
          }
        }
      },
      "span": {
        "start": 387,
        "end": 409
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "generics"
            },
            "span": {
              "start": 413,
              "end": 421
            }
          },
          "params": {
            "Multi": [
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "a"
                      },
                      "span": {
                        "start": 431,
                        "end": 432
                      }
                    },
                    {
                      "start": 431,
                      "end": 432
                    }
                  ]
                },
                "ty": {
                  "Reference": {
                    "name": {
                      "symbol": {
                        "unstable_source": "T"
                      },
                      "span": {
                        "start": 434,
                        "end": 435
                      }
                    },
                    "arguments": null,
                    "span": {
                      "start": 434,
                      "end": 435
                    }
                  }
                },
                "id": 0,
                "span": {
                  "start": 431,
                  "end": 435
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "b"
                      },
                      "span": {
                        "start": 437,
                        "end": 438
                      }
                    },
                    {
                      "start": 437,
                      "end": 438
                    }
                  ]
                },
                "ty": {
                  "Reference": {
                    "name": {
                      "symbol": {
                        "unstable_source": "U"
                      },
                      "span": {
                        "start": 440,
                        "end": 441
                      }
                    },
                    "arguments": null,
                    "span": {
                      "start": 440,
                      "end": 441
                    }
                  }
                },
                "id": 0,
                "span": {
                  "start": 437,
                  "end": 441
                }
              }
            ]
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 447,
              "end": 449
            }
          },
          "return_ty": {
            "Reference": {
              "name": {
                "symbol": {
                  "unstable_source": "V"
                },
                "span": {
                  "start": 445,
                  "end": 446
                }
              },
              "arguments": null,
              "span": {
                "start": 445,
                "end": 446
              }
            }
          },
          "is_async": false,
          "generics": {
            "params": [
              {
                "symbol": {
                  "unstable_source": "T"
                },
                "span": {
                  "start": 422,
                  "end": 423
                }
              },
              {
                "symbol": {
                  "unstable_source": "U"
                },
                "span": {
                  "start": 425,
                  "end": 426
                }
              },
              {
                "symbol": {
                  "unstable_source": "V"
                },
                "span": {
                  "start": 428,
                  "end": 429
                }
              }
            ]
          },
          "span": {
            "start": 410,
            "end": 449
          }
        }
      },
      "span": {
        "start": 410,
        "end": 449
      }
    },
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "mixed_generics"
            },
            "span": {
              "start": 453,
              "end": 467
            }
          },
          "params": {
            "Multi": [
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "a"
                      },
                      "span": {
                        "start": 474,
                        "end": 475
                      }
                    },
                    {
                      "start": 474,
                      "end": 475
                    }
                  ]
                },
                "ty": {
                  "Reference": {
                    "name": {
                      "symbol": {
                        "unstable_source": "T"
                      },
                      "span": {
                        "start": 477,
                        "end": 478
                      }
                    },
                    "arguments": null,
                    "span": {
                      "start": 477,
                      "end": 478
                    }
                  }
                },
                "id": 0,
                "span": {
                  "start": 474,
                  "end": 478
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "b"
                      },
                      "span": {
                        "start": 480,
                        "end": 481
                      }
                    },
                    {
                      "start": 480,
                      "end": 481
                    }
                  ]
                },
                "ty": {
                  "String": {
                    "start": 483,
                    "end": 489
                  }
                },
                "id": 0,
                "span": {
                  "start": 480,
                  "end": 489
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "symbol": {
                        "unstable_source": "c"
                      },
                      "span": {
                        "start": 491,
                        "end": 492
                      }
                    },
                    {
                      "start": 491,
                      "end": 492
                    }
                  ]
                },
                "ty": {
                  "Reference": {
                    "name": {
                      "symbol": {
                        "unstable_source": "U"
                      },
                      "span": {
                        "start": 494,
                        "end": 495
                      }
                    },
                    "arguments": null,
                    "span": {
                      "start": 494,
                      "end": 495
                    }
                  }
                },
                "id": 0,
                "span": {
                  "start": 491,
                  "end": 495
                }
              }
            ]
          },
          "body": {
            "stmts": [],
            "id": 0,
            "span": {
              "start": 501,
              "end": 503
            }
          },
          "return_ty": {
            "Reference": {
              "name": {
                "symbol": {
                  "unstable_source": "U"
                },
                "span": {
                  "start": 499,
                  "end": 500
                }
              },
              "arguments": null,
              "span": {
                "start": 499,
                "end": 500
              }
            }
          },
          "is_async": false,
          "generics": {
            "params": [
              {
                "symbol": {
                  "unstable_source": "T"
                },
                "span": {
                  "start": 468,
                  "end": 469
                }
              },
              {
                "symbol": {
                  "unstable_source": "U"
                },
                "span": {
                  "start": 471,
                  "end": 472
                }
              }
            ]
          },
          "span": {
            "start": 450,
            "end": 503
          }
        }
      },
      "span": {
        "start": 450,
        "end": 503
      }
    }
  ]
}
